### PR TITLE
rewind 10161.1 (new cask)

### PR DESCRIPTION
**Pull request by razvanazamfirei**
**Important:** *Do not tick a checkbox if you haven’t performed its action.* Honesty is indispensable for a smooth review process.

_In the following questions `<cask>` is the token of the cask you're submitting._

After making any changes to a cask, existing or new, verify:

- [x] The submission is for [a stable version](https://docs.brew.sh/Acceptable-Casks#stable-versions) or [documented exception](https://docs.brew.sh/Acceptable-Casks#but-there-is-no-stable-version).
- [x] `brew audit --cask --online <cask>` is error-free.
- [x] `brew style --fix <cask>` reports no offenses.

Additionally, **if adding a new cask**:

- [x] Named the cask according to the [token reference](https://docs.brew.sh/Cask-Cookbook#token-reference).
- [x] Checked the cask was not [already refused](https://github.com/Homebrew/homebrew-cask/search?q=is%3Aclosed&type=Issues).
- [x] Checked the cask is submitted to [the correct repo](https://docs.brew.sh/Acceptable-Casks#finding-a-home-for-your-cask).
- [x] `brew audit --new-cask <cask>` worked successfully.
- [x] `brew install --cask <cask>` worked successfully.
- [x] `brew uninstall --cask <cask>` worked successfully.

---

This was originally submitted in #143139 by @nihaals. I wrote the livecheck to update every 3 days; it looks like they publish new versions every few hours or so. Would appreciate a second set of eyes on it. I know a versioned url exists, but it includes a random string that isn't found anywhere else in the appcast which is why I opted to use an unversioned link.

Note: This binary is universal, but the app only works on arm. CI skipped the installation process (presumably due to the architecture limitation), but it works locally.

### CODE REVIEW

**Comment by nihaals** [comment]
Are there any other casks that follow this? It seems like it essentially replaces `:latest` in some cases and I think I would personally be in favour of using that instead as it seems more accurate to the situation (and probably just restoring the previous PR since that already matches this). `:latest` already suggests that the cask might be updated too frequently and is reserved for special cases, while this requires reading the actual cask to see what the version means and why it's out of date/`livecheck` doesn't give the expected output.

As for the versioned URLs, we could parse the commit SHA from the URL and make it a CSV version with the normal version number and the commit. I can also confirm that even though the appcast only includes a limited number of versions, the previous versions still exist, but I'm not sure how viable this would be to use as there would be a stronger incentive to keep the version as up to date as possible.

**Comment by razvanazamfirei** [comment]
> Are there any other casks that follow this? 

Having a livecheck rate? Yes. I can think of `grandtotal` and `bettertouchtool` off the top of my head. 
> It seems like it essentially replaces :latest in some cases and I think I would personally be in favour of using that instead

The tradeoff is that casks with `:latest` do not get upgraded by default when running `brew upgrade`, you'd need to either run `brew upgrade --greedy` or `brew upgrade --greedy-latest`. That being said, the app also looks like it auto-updates, which would also not be included by default in `brew upgrade`. I guess really the question is whether there's a need to have this versioned at all. Happy to move forward with `:latest`, with the caveat that the cask will be useful for initially installing the application, but not to keep it update.

File: Casks/rewind.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  version "1.0041,20230419"
+  sha256 :no_check
+
+  url "https://download.rewind.ai/Rewind.dmg"
+  name "Rewind"
+  desc "Record and search your screen and audio"
+  homepage "https://www.rewind.ai/"
+
+  livecheck do
+    url "https://updates.rewind.ai/appcasts/main.xml"
```
Comment (reitermarkus) [inline Casks/rewind.rb:11]: Can this not still use `:sparkle`?

**Comment by reitermarkus** [comment]
We want to use versions wherever possible instead of `:latest`.

**Comment by reitermarkus** [comment]
> As for the versioned URLs, we could parse the commit SHA from the URL

Yes, if there is a versioned URL we should use that.

File: Casks/rewind.rb (hunk 1 above)
Comment (samford) [reply Casks/rewind.rb:11]: It can, and it simplifies the `strategy` block quite a bit. I'll add a suggestion in a moment.

File: Casks/rewind.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    strategy :xml do |xml|
+      # Throttle updates to one every 3 days.
+      next version if DateTime.parse(version.csv.second) + 3 > Date.today
+
+      item = xml.get_elements("//item")[0]
+      version = item.elements["sparkle:shortVersionString"].text
+      date = DateTime.parse(item.elements["pubDate"].text).strftime("%Y%m%d")
+
+      "#{version},#{date}"
```
Comment (samford) [inline Casks/rewind.rb:20]:
```suggestion
    strategy :sparkle do |item|
      # Throttle updates to one every 3 days.
      next version if DateTime.parse(version.csv.second) + 3 > Date.today

      "#{item.short_version},#{item.pub_date.strftime("%Y%m%d")}"
```

As Markus mentioned, we can use the `Sparkle` strategy here. It takes care of most of the custom logic (e.g., getting the first item, pulling out the `shortVersionString` text, and parsing the `pubDate` into a `Time`/`DateTime` object), so the `strategy` block ends up pretty simple.

**Review by samford — COMMENTED** [review]
For what it's worth, throttling versions in a `livecheck` block is just a temporary, stop-gap measure until someone implements a proper solution in related tooling. I've always been of the opinion that livecheck should return the latest [appropriate] version and any version throttling should happen outside of livecheck.

However, with casks that update very frequently, it was either throttle versions in `livecheck` blocks (for now) or require maintainers to continue to deal with the onslaught of version bump PRs (some were updating multiple times a day), so here we are. [For reference, we had an internal discussion about this on 2023-03-01 in the `#cask` Slack channel.]

File: Casks/rewind.rb (hunk 2 above)
Comment (razvanazamfirei) [reply Casks/rewind.rb:20]: Thank you! This is super helpful. At some point, I'll update the documentation in the Caskbook, so other people don't try to reinvent the wheel in the future. :)

File: Casks/rewind.rb [hunk 3]
```diff
@@ -1,8 +1,8 @@
 cask "rewind" do
-  version "1.0041,20230419"
-  sha256 :no_check
+  version "10056.1,2c3fbe8,20230420"
```
Comment (nihaals) [inline Casks/rewind.rb:2]: Should we be matching the version visible to the user (`1.0056`)?

Comment (razvanazamfirei) [reply Casks/rewind.rb:2]: The url uses 10056.1, which is why I matched on that. `https://updates.rewind.ai/builds/main/b10056.1-main-2c3fbe8.zip`

We could change the version back to the short one and do something silly like: 
```
#{version.csv.first.gsub(/(\d)\.(\d{4})/, '\1\2.1')}
```
but I don't know if it's worth it.

**Comment by reitermarkus** [comment]
> A bit of an aside but are there any plans for auto-merging simple bump PRs (i.e. only updating version and checksum)?

We used to do just that. It resulted in a security incident, so we won't do that again.

**Review by reitermarkus — APPROVED** [review]

**Comment by nihaals** [comment]
> > A bit of an aside but are there any plans for auto-merging simple bump PRs (i.e. only updating version and checksum)?
> 
> We used to do just that. It resulted in a security incident, so we won't do that again.

@reitermarkus Is there an issue/discussion/write up of this somewhere? I did think about security but couldn't think of a way to exploit it vs. the current system so dismissed it and would be interested to see more about it.

**Comment by SMillerDev** [comment]
https://brew.sh/2021/04/21/security-incident-disclosure/

**Pull request merged**